sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/StructFilters.scala

```diff
@@ -0,0 +1,161 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst
+
+import scala.util.Try
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.sources
+import org.apache.spark.sql.types.{BooleanType, StructType}
+
+/**
+ * The class provides API for applying pushed down filters to partially or
+ * fully set internal rows that have the struct schema.
+ *
+ * @param filters The pushed down source filters. The filters should refer to
+ *                the fields of the provided schema.
+ * @param schema The required schema of records from datasource files.
```

Comment:
Actually, the previous `requiredSchema` was more intuitive, but this one also good in this context.

Reply:
The term `requiredSchema` makes sense in CSV datasource because it operates on 2 schemas: `dataSchema` is full schema of CSV records, and `requiredSchema` is requested by the upper layer. JSON datasource operates on only one schema which is `requiredSchema` in terms of CSV. JSON datatsource can do that because it parses input strings in streaming way, and can skip JSON fields are not presented the schema. Comparing to JSON, CSV parses entire input that's why it needs full schema.

`StructFilters` needs a schema to bind filters attributes to actual position in a row. required or not required schema doesn't matter for `StructFilters`.

Reply:
Yep. I agree.